File: src/test/java/Theories.java
import fortress.tfol.*;

import java.util.List;

public class Theories {
    
    // Transitive relation R
    // Anti symmetric
    // Anti reflexive
    // BiggerFish: For every x, there exists y such that x R y
    // Should be satisfiable only for infinite domains, or empty domains
    // public static Theory lessThanInfinite = makeLessThanInfinite();
    
    
    
    
    
    
    // Group theory axioms
    
    static Sort G = Sort.mkSortConst("G"); // group type
    static Var e = Term.mkVar("e"); // identity element
    static FuncDecl f = FuncDecl.mkFuncDecl("f", G, G, G); // group operation
    
    public static Theory groupTheory = constructGroupTheory();
    public static Theory nonAbelianGroupTheory = constructNonAbelianGroupTheory();
    
    private static Theory constructGroupTheory() {
        Var x = Term.mkVar("x");
        Var y = Term.mkVar("y");
        Var z = Term.mkVar("z");
        
        Term associativeAxiom = Term.mkForall(List.of(x.of(G), y.of(G), z.of(G)),
            Term.mkEq(
                Term.mkApp("f", Term.mkApp("f", x, y), z),
                Term.mkApp("f", x, Term.mkApp("f", y, z))));
        
        Term identityAxiom = Term.mkForall(x.of(G),
            Term.mkAnd(
                Term.mkEq(Term.mkApp("f", x, e), x),
                Term.mkEq(Term.mkApp("f", e, x), x)));
        
        Term inverseAxiom = Term.mkForall(x.of(G), Term.mkExists(y.of(G), 
            Term.mkAnd(
                Term.mkEq(Term.mkApp("f", x, y), e),
                Term.mkEq(Term.mkApp("f", y, x), e))));
                
        return Theory.empty()
            .withSort(G)
            .withConstant(e.of(G))
            .withFunctionDeclaration(f)
            .withAxiom(associativeAxiom)
            .withAxiom(identityAxiom)
            .withAxiom(inverseAxiom);
    }
    
    public static Theory constructNonAbelianGroupTheory() {
        Var x = Term.mkVar("x");
        Var y = Term.mkVar("y");
        Term abelianAssertion = Term.mkForall(List.of(x.of(G), y.of(G)),
            Term.mkEq(
                Term.mkApp("f", x, y),
                Term.mkApp("f", y, x)));
        return groupTheory.withAxiom(Term.mkNot(abelianAssertion));
    }
}

File: core/src/test/scala/Theories.scala
import fortress.msfol._

object Theories {
    
    // Transitive relation R
    // Anti symmetric
    // Anti reflexive
    // BiggerFish: For every x, there exists y such that x R y
    // Should be satisfiable only for infinite domains, or empty domains
    // public static Theory lessThanInfinite = makeLessThanInfinite()
    
    // Group theory axioms
    
    val G = Sort.mkSortConst("G") // group type
    val e = Var("e") // identity element
    val f = FuncDecl.mkFuncDecl("f", G, G, G) // group operation
    
    val groupTheory: Theory = {
        val x = Var("x")
        val y = Var("y")
        val z = Var("z")
        
        val associativeAxiom = Forall(Seq(x.of(G), y.of(G), z.of(G)),
            Eq(
                App("f", App("f", x, y), z),
                App("f", x, App("f", y, z))))
        
        val identityAxiom = Forall(x.of(G),
            And(
                Eq(App("f", x, e), x),
                Eq(App("f", e, x), x)))
        
        val inverseAxiom = Forall(x.of(G), Exists(y.of(G), 
            And(
                Eq(App("f", x, y), e),
                Eq(App("f", y, x), e))))
                
        Theory.empty
            .withSort(G)
            .withConstantDeclaration(e.of(G))
            .withFunctionDeclaration(f)
            .withAxiom(associativeAxiom)
            .withAxiom(identityAxiom)
            .withAxiom(inverseAxiom)
    }
    
    val nonAbelianGroupTheory: Theory = {
        val x = Var("x")
        val y = Var("y")
        val abelianAssertion = Forall(Seq(x.of(G), y.of(G)),
            Eq(
                App("f", x, y),
                App("f", y, x)))
        groupTheory.withAxiom(Not(abelianAssertion))
    }
}
